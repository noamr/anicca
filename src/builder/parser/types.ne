@{%
const moo = require("moo")
const lexer = moo.compile({
  keywords: ['i8', 'u8', 'i16', 'u16', 'i32', 'u32', 'i64', 'u64', 'f32', 'f64', 'i128', 'u128', 'string', 'bool', 'ByteArray'],
  ws: /[ \t]+/,
  word: /[A-Za-z_$][A-Za-z_$0-9]*/,
  symbols: /[\[\],{}:]/
})

%}
@builtin "whitespace.ne"

@lexer lexer

type ->
    singular {% id %}
    | tuple {% id %}
    | dictionary {% id %}

singular ->
    _ %keywords _ {% ([,{value}]) => value %}

tuple ->
    _ "[" _ args _ "]" _ {% ([,,,args]) => ({tuple: args}) %}

args ->
    type
    | _ args _ "," _ type _ {% ([,multi,,,, single]) => [...multi, single] %}

dictionary ->
    _ "{" _ type _ ":" _ type "}" {% ([,,,a,,,,b]) => ({dictionary: [a, b]}) %}