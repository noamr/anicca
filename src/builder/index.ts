import fs from 'fs'
import path from 'path'
import nearley from 'nearley'
import YAML from 'yaml'
import { parseKal, ParseOptions } from './parser/index'
import { removeUndefined } from './transformers/helpers'
import {Bundle} from './types'
import transformBundle from './transformers/transform'
const {rollup} = require('rollup')
const rollupJson = require('rollup-plugin-json')
const rollupTypescript = require('rollup-plugin-typescript')
import sourcemaps from 'rollup-plugin-sourcemaps'

interface BuildOptions {
    inputPath?: string
    src?: string
    outputDir: string
    rollupConfig?: any
}

export function parse(yamlString: string, opt: ParseOptions = {internal: false}): Bundle {
    return removeUndefined(parseKal(YAML.parse(yamlString), opt))
}

export async function build(config: BuildOptions): Promise<void> {
    const bundle = parse(config.src || fs.readFileSync(config.inputPath || '', 'utf8'))
    const {store, views, channels, routes, headers} = transformBundle(bundle)
    const toOutputPath = (p: string) => path.resolve(config.outputDir, p)
    const resolveLib = (p: string) => path.relative(config.outputDir, path.resolve(__dirname, p))

    const storeOutputPath = toOutputPath('store.json')
    const viewsOutputPath = toOutputPath('views.json')
    const routesOutputPath = toOutputPath('routes.json')
    const busOutputPath = toOutputPath('channels.json')
    const mainOutputPath = toOutputPath('main.js')
    const workerOutputPath = toOutputPath('worker.js')
    const mainWrapperOutputPath = toOutputPath('main-wrapper.js')
    const workerWrapperOutputPath = toOutputPath('worker-wrapper.js')
    const headersOutputPath = toOutputPath('headers.json')

    fs.writeFileSync(busOutputPath, JSON.stringify(channels, null, 4))
    fs.writeFileSync(storeOutputPath, JSON.stringify(store, null, 4))
    fs.writeFileSync(viewsOutputPath, JSON.stringify(views, null, 4))
    fs.writeFileSync(routesOutputPath, JSON.stringify(routes, null, 4))
    fs.writeFileSync(headersOutputPath, JSON.stringify(headers, null, 4))

    const mainWrapper = `
        import main from '${resolveLib('../runtime/common/main')}'
        import views from './views.json'
        import channels from './channels.json'

        export default function init({rootElements, routes}) {
            return main({rootElements, routes, views, channels, storeWorkerPath: 'worker.js'})
        }
    `

    const interpreterWorkerWrapper = `
        import initStore from '${resolveLib('../runtime/interpreter/ShellWorker')}'
        import store from './store.json'

        initStore(store)
    `

    const rollupConfig = (name: string) => ({
        inputOptions: {
            input: `${toOutputPath(name)}-wrapper.js`,
            plugins: [
                rollupTypescript(),
                rollupJson({indent: '    ', preferConst: true, compact: true, namedExports: false}),
                sourcemaps()
            ]
        },
        outputOptions: {
            file: `${toOutputPath(name)}.js`,
            format: 'esm'
        },
    })

    fs.writeFileSync(workerWrapperOutputPath, interpreterWorkerWrapper)
    fs.writeFileSync(mainWrapperOutputPath, mainWrapper)

    const mainConfig = rollupConfig('main')
    const workerConfig = rollupConfig('worker')
    const b1 = await rollup(mainConfig.inputOptions)
    const b2 = await rollup(rollupConfig('worker').inputOptions)
    await b1.write(mainConfig.outputOptions)
    await b2.write(workerConfig.outputOptions)
}
