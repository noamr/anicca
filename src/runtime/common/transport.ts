import { NativeType, NativeDictionaryType, NativeTupleType } from '../../builder/types'

function encodeNumber(value: number, f: keyof DataView, size: number): ArrayBuffer[] {
    const dv = new DataView(new ArrayBuffer(size));
    (dv[f] as ((a: number, b: number) => void))(0, value)
    return [dv.buffer]
}
function encodeSingle(value: any, type: NativeType): ArrayBuffer[] {
    switch (type) {
        case 'null':
            return []
        case 'u32':
            return encodeNumber(value, 'setUint32', 4)
        case 'i32':
            return encodeNumber(value, 'setInt32', 4)
        case 'u16':
            return encodeNumber(value, 'setUint16', 2)
        case 'i16':
            return encodeNumber(value, 'setInt16', 2)
        case 'u8':
            return encodeNumber(value, 'setUint8', 1)
        case 'i8':
            return encodeNumber(value, 'setInt8', 1)
        case 'f32':
            return encodeNumber(value, 'setFloat32', 4)
        case 'f64':
        case 'u64':
        case 'i64':
            return encodeNumber(value, 'setFloat64', 8)
        case 'bool':
            return encodeNumber(value ? 1 : 0, 'setUint8', 1)
        case 'string': {
            const {buffer} = new TextEncoder().encode(value as string)
            return [...encodeSingle(buffer.byteLength, 'u32'), buffer as ArrayBuffer]
        }
        case 'ByteArray': {
            if (typeof value !== 'object' || !(value instanceof ArrayBuffer))
                throw new Error(`Expecting ArrayBuffer when encoding ${value}`)
            return [...encodeSingle((value as ArrayBuffer).byteLength, 'u32'), value as ArrayBuffer]
        }
        default:
            throw new Error(`Unknown native type: ${type}`)
    }
}

function decodeSingle(view: DataView, type: NativeType): [any, DataView] {
    switch (type) {
        case 'u32':
            return [view.getUint32(0), new DataView(view.buffer, view.byteOffset + 4)]
        case 'i32':
            return [view.getInt32(0), new DataView(view.buffer, view.byteOffset + 4)]
        case 'u16':
            return [view.getUint16(0), new DataView(view.buffer, view.byteOffset + 2)]
        case 'i16':
            return [view.getInt16(0), new DataView(view.buffer, view.byteOffset + 2)]
        case 'u8':
            return [view.getUint8(0), new DataView(view.buffer, view.byteOffset + 1)]
        case 'i8':
            return [view.getInt8(0), new DataView(view.buffer, view.byteOffset + 1)]
        case 'f32':
            return [view.getFloat32(0), new DataView(view.buffer, view.byteOffset + 4)]
        case 'u64':
        case 'i64':
        case 'f64':
            return [view.getFloat64(0), new DataView(view.buffer, view.byteOffset + 8)]
        case 'bool':
            return [!!view.getUint8(0), new DataView(view.buffer, view.byteOffset + 1)]
        case 'string': {
            const length = view.getUint32(0)
            return [new TextDecoder().decode(new DataView(view.buffer, view.byteOffset + 4, length)),
                    new DataView(view.buffer, view.byteOffset + 4 + length)]
        }
        case 'ByteArray': {
            const size = view.getUint32(0)
            return [view.buffer.slice(view.byteOffset + 4, view.byteOffset + 4 + size),
                new DataView(view.buffer, view.byteOffset + 4 + size)]
        }
        default:
            throw new Error(`Unknown native type: ${type}`)
    }
}

function decodeAny(buffer: DataView, type: NativeType): [any, DataView] {
    if (type === 'null')
        return [null, new DataView(new ArrayBuffer(0))]
    if (typeof type === 'string')
        return decodeSingle(buffer, type)

    if (Reflect.has(type, 'dictionary')) {
        const {dictionary} = type as NativeDictionaryType
        return decodeDictionary(buffer, dictionary[0], dictionary[1])
    }

    if (Reflect.has(type, 'tuple')) {
        const {tuple} = type as NativeTupleType<any>
        return decodeTuple(buffer, tuple)
    }

    throw new Error(`Unknown type: ${type}`)
}

function decodeDictionary(view: DataView, keyType: NativeType, valueType: NativeType): [Map<any, any>, DataView] {
    const size = view.getUint32(0)
    view = new DataView(view.buffer, view.byteOffset + 4)
    const entries = Array(size).fill(null).map((n, i) => {
        const keyResult = decodeAny(view, keyType)
        view = keyResult[1]
        const valueResult = decodeAny(view, valueType)
        view = valueResult[1]
        return [keyResult[0], valueResult[0]]
    }) as [[any, any]]

    return [new Map(entries), view]
}

function decodeTuple(view: DataView, types: NativeType[]): [any[], DataView] {
    const results = new Array(types.length).fill(null).map((n, i) => {
        const result = decodeAny(view, types[i])
        view = result[1]
        return result[0]
    }) as any[]

    return [results, view]
}

function encodeDictionary(value: Map<any, any>, keyType: NativeType, valueType: NativeType): ArrayBuffer[] {
    return [
        ...encodeAny(value.size, 'u32'),
        ...[...value.entries()].flatMap(([key, value]) => [...encodeAny(key, keyType), ...encodeAny(value, valueType)])
    ]
}

function encodeAny(value: any, type: NativeType): ArrayBuffer[] {
    if (typeof type === 'string')
        return encodeSingle(value, type)

    if (Reflect.has(type, 'dictionary')) {
        const [keyType, valueType] = (type as NativeDictionaryType).dictionary
        if (typeof value !== 'object') {
            throw new Error(`Expecting object when encoding ${value}`)
        }

        if (!(value instanceof Map)) {
            throw new Error(`Expecting map when encoding ${value}`)
        }

        return encodeDictionary(value as Map<any, any>, keyType, valueType)
    }

    if (Reflect.has(type, 'tuple')) {
        const tt = type as NativeTupleType<any>

        if (!Array.isArray(value))
            throw new Error(`Expecting array when encoing ${value}`)
        const a = value as any[]
        if (a.length !== tt.tuple.length)
            throw new Error(`Tuple length of ${value} is ${value.length}, expected ${tt.tuple.length}`)

        return a.flatMap((v, i) => encodeAny(v, tt.tuple[i]))
    }

    throw new Error(`Unknown value: ${value}`)
}

export function encode(value: any, type: NativeType): ArrayBuffer {
    const buffers = encodeAny(value, type)
    const result = new ArrayBuffer(buffers.reduce((a, b) => a + b.byteLength, 0))
    const u8 = new Uint8Array(result)
    let offset = 0
    buffers.forEach(b => {
        u8.set(new Uint8Array(b), offset)
        offset += b.byteLength
    })

    return result
}

export function decode(buffer: ArrayBuffer, type: NativeType): any {
    return buffer ? decodeAny(new DataView(buffer), type)[0] : null
}