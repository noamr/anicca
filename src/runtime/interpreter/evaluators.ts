import { NativeType } from '../../builder/types'

export interface Context<K = any, V = any, A = any> {
    key: K
    source: Evaluator<V>
    aggregate?: A
}

export type Evaluator<T = any> = ((context: Context | null) => T) & {
    type?: NativeType
    token?: any
    originalFormula?: Formula
}

const u2n = (a: any) => typeof a === 'undefined' ? null : a

const extractMathFunctions = (...keys: Array<keyof Math>) =>
    keys.map(k => ({k: (...args: Evaluator[]) => (ctx: Context) =>
        (Math[k] as any).apply(null, args.map(a => a(ctx)))})).reduce((a, o) => Object.assign(a, o), {})

const extractStringFunctions = <T extends keyof ''>(...keys: T[]) =>
    keys.map(k => ({k: (s: Evaluator, ...args: Evaluator[]) => (ctx: Context) =>
        ((''[k]) as any).call(s(ctx), ...args.map(a => a(ctx)))})).reduce((a, o) => Object.assign(a, o), {})

const flatMap = <K, V, K2, V2>(source: Evaluator<Map<K, V>>, predicate: Evaluator<Map<number, Map<number, K2|V2>>>) =>
    (ctx: Context|null) => new Map(Array.from(source(ctx).keys()).flatMap(key => 
        Array.from(predicate({ source, key }))))

const flatReduce = (src: Evaluator, predicate: Evaluator, initialValue: Evaluator) => (ctx: Context|null) => {
    const source = src(ctx)
    const keys = [...source.keys()]
    let aggregate = initialValue(ctx)
    for (const key of keys) {
        const [result, final] = predicate({source, key, aggregate})
        if (final)
            return result

        aggregate = result
    }

    return aggregate
}

export const defaultEvaluators: {[op: string]: (...args: Evaluator[]) => Evaluator} = {
    // operators
    plus: (a, b) => c => a(c) + b(c),
    minus: (a, b) => c => a(c) - b(c),
    mult: (a, b) => c => a(c) * b(c),
    div: (a, b) => c => a(c) / b(c),
    mod: (a, b) => c => a(c) % b(c),
    pow: (a, b) => ctx => a(ctx) ** b(ctx),
    negate: a => ctx => -(a(ctx)),
    shl: (a, b) => ctx => a(ctx) << b(ctx),
    shr: (a, b) => ctx => a(ctx) >> b(ctx),
    ushr: (a, b) => ctx => a(ctx) >>> b(ctx),
    bwnot: a => ctx => ~(a(ctx)),
    bwand: (a, b) => ctx => a(ctx) & b(ctx),
    bwor: (a, b) => ctx => a(ctx) | b(ctx),
    bwxor: (a, b) => ctx => a(ctx) ^ b(ctx),

    // comparison
    eq: (a, b) => ctx => a(ctx) === b(ctx),
    neq: (a, b) => ctx => a(ctx) !== b(ctx),
    lt: (a, b) => ctx => a(ctx) < b(ctx),
    gt: (a, b) => ctx => a(ctx) > b(ctx),
    gte: (a, b) => ctx => a(ctx) >= b(ctx),
    lte: (a, b) => ctx => a(ctx) <= b(ctx),

    // Casting
    parseInt: (s, r) => ctx => parseInt(s(ctx), r(ctx)),
    parseFloat: (s) => ctx => parseFloat(s(ctx)),
    formatFloat: (n, f) => ctx => Number(n(ctx)).toFixed(f(ctx)),
    formatInt: (n, r) => ctx => Number(n(ctx)).toString(r(ctx)),

    // Math
    max: (a, b) => (ctx) => Math.max(a(ctx), b(ctx)),
    min: (a, b) => (ctx) => Math.min(a(ctx), b(ctx)),
    sin: a => ctx => Math.sin(a(ctx)),
    cos: a => ctx => Math.cos(a(ctx)),
    asin: a => ctx => Math.asin(a(ctx)),
    acos: a => ctx => Math.acos(a(ctx)),
    tan: a => ctx => Math.tan(a(ctx)),
    atan: a => ctx => Math.atan(a(ctx)),
    abs: a => ctx => Math.abs(a(ctx)),
    floor: a => ctx => Math.floor(a(ctx)),
    ceil: a => ctx => Math.ceil(a(ctx)),
    round: a => ctx => Math.round(a(ctx)),
    random: () => () => Math.random(),

    // String
    toLowerCase: a => ctx => (a(ctx) as string).toLowerCase(),
    toUpperCase: a => ctx => (a(ctx) as string).toUpperCase(),
    charAt: (a, b) => ctx => (a(ctx) as string).charAt(b(ctx)),
    startsWith: (a, b) => ctx => (a(ctx) as string).startsWith(a(ctx)),
    endsWith: (a, b) => ctx => (a(ctx) as string).endsWith(b(ctx)),
    stringIncludes: (a, b) => ctx => (a(ctx) as string).includes(b(ctx)),
    trim: (a) => ctx => (a(ctx) as string).trim(),
    match: (a, b) => ctx => (a(ctx) as string).match(b(ctx)),
    join: (args, separator) => ctx => [...args(ctx).values()].join(separator(ctx)),

    // Logical
    isNil: a => ctx => a(ctx) === null,
    cond: (a, b, c) => ctx => a(ctx) ? b(ctx) : c(ctx),
    not: a => ctx => !(a(ctx)),

    // Constructing
    tuple: <A, B>(...args: Array<Evaluator<A>>): Evaluator<A[]> => ctx => args.map(a => a(ctx)),
    array: <A>(...entries: Array<Evaluator<A>>): Evaluator<Map<number, A>> => ctx =>
        new Map(entries.map((a, i) => [i, a(ctx)] as [number, A])),
    object: <K, V>(...entries: Array<Evaluator<[K, V]>>): Evaluator<Map<K, V>> => ctx =>
        new Map(entries.map(e => e(ctx))),

    // Maps
    get: (a, b) => ctx => u2n((obj => obj && obj.get(b(ctx)))(a(ctx))),
    at: (a, b) => ctx => u2n((obj => obj && obj[b(ctx)])(a(ctx))),
    size: a => ctx => (a(ctx) || {size: 0}).size,
    head: a => ctx => u2n([...(a(ctx) as Map<any, any>).keys()][0]),
    tail: a => ctx => u2n([...(a(ctx) as Map<any, any>).keys()].slice(-1)[0]),
    first: a => ctx => a(ctx)[0],
    second: a => ctx => a(ctx)[1],
    flatMap,
    flatReduce,

    // contextual
    key: () => ctx => ctx && ctx.key,
    value: () => ctx => ctx && ctx.source(ctx).get(ctx.key),
    aggregate: () => ctx => ctx && ctx.aggregate,
    source: () => ctx => ctx && ctx.source(ctx),

    // other
    now: () => () => Date.now()
}
