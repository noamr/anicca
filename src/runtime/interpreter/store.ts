import { Enqueue, StartStore } from '../shell/RuntimeTypes'
import { StoreDefinition, Slot } from '../../builder/StoreDefinition'

export default function createStore(def: StoreDefinition) : StartStore {
    return function initStore(dequeue: Enqueue): Enqueue {
        const tables = def.tableTypes.map(t => ({}))
        let memos: {[key: number]: any} = {}
        const slots = def.slots.map(resolveSlot)
        const nextID = (n => () => ++n)(0)
        const extractMathFunctions = (...keys: (keyof Math)[]) =>
            keys.map(k => ({k: (...args: Evaluator[]) => (ctx: Context) => (Math[k] as any).apply(null, args.map(a => a(ctx)))})).reduce((a, o) => Object.assign(a, o), {})

        const extractStringFunctions = <T extends keyof ''>(...keys: T[]) =>
            keys.map(k => ({k: (s: Evaluator, ...args: Evaluator[]) => (ctx: Context) => ((''[k]) as any).call(s(ctx), ...args.map(a => a(ctx)))})).reduce((a, o) => Object.assign(a, o), {})

        const tMap = (table: Evaluator, predicate: Evaluator) =>
                (ctx: Context|null) => Object.keys(tables[table(ctx)]).map((key: string|number) => predicate({key, source: table, aggregate: null}))
                    .reduce((a: any, b: any) => ({...a, ...b}), {})
    
        const tReduce = (table: Evaluator, predicate: Evaluator, initialValue: Evaluator) =>
            (ctx: Context|null) => Object.keys(table).reduce((aggregate: any, key: any) =>
                predicate({source: table(ctx), key, aggregate}), initialValue(ctx))
    
        type Context = {
            key: string | number
            source: any
            aggregate: any
        }

        type Evaluator = (context: Context | null) => any

        const resolvers : {[op: string]: (...args: Evaluator[]) => Evaluator} = {
            plus: (a, b) => c => a(c) + b(c),
            minus: (a, b) => c => a(c) - b(c),
            mult: (a, b) => c => a(c) * b(c),
            div: (a, b) => c => a(c) / b(c),
            mod: (a, b) => c => a(c) % b(c),
            cond: (a, b, c) => ctx => a(ctx) ? b(ctx) : c(ctx),
            isnil: a => ctx => (v => (v === null || typeof v === 'undefined'))(a(ctx)),
            array: (...args: Evaluator[]) => ctx => args.map(a => a(ctx)),
            not: a => ctx => !(a(ctx)),
            bwnot: a => ctx => ~(a(ctx)),
            negate: a => ctx => -(a(ctx)),
            pow: (a, b) => ctx => a(ctx) ** b(ctx),
            get: (a, b) => ctx => a(ctx)[b(ctx)],
            shl: (a, b) => ctx => a(ctx) << b(ctx),
            shr: (a, b) => ctx => a(ctx) >> b(ctx),
            ushr: (a, b) => ctx => a(ctx) >>> b(ctx),
            eq: (a, b) => ctx => a(ctx) === b(ctx),
            lt: (a, b) => ctx => a(ctx) < b(ctx),
            gt: (a, b) => ctx => a(ctx) > b(ctx),
            gte: (a, b) => ctx => a(ctx) >= b(ctx),
            lte: (a, b) => ctx => a(ctx) <= b(ctx),
            bwand: (a, b) => ctx => a(ctx) & b(ctx),
            bwor: (a, b) => ctx => a(ctx) | b(ctx),
            bwxor: (a, b) => ctx => a(ctx) ^ b(ctx),
            and: (a, b) => ctx => a(ctx) && b(ctx),
            or: (a, b) => ctx => a(ctx) || b(ctx),
            entry: (a, b) => ctx => ({[a(ctx)]: b(ctx)}),
            key: () => ctx => ctx && ctx.key,
            primitive: a => ctx => def.primitives[a(ctx)],
            table: (a) => ctx => tables[a(ctx)], 
            value: () => ctx => ctx && ctx.source[ctx.key],
            aggregate: () => ctx => ctx && ctx.aggregate,
            source: () => ctx => ctx && ctx.source,
            now: () => () => Date.now(),
            ref: (slot) => ctx => evaluateMemoized(slot(ctx), ctx),
            object: (...entries: Evaluator[]) => ctx => entries.reduce((a, e) => Object.assign(a, e(ctx)), {}),
            ...extractMathFunctions('sin', 'cos', 'max', 'log', 'random', 'log2', 'log10', 'tan', 'acos', 'asin', 'sqrt', 'floor', 'ceil', 'trunc'),
            ...extractStringFunctions('toLowerCase', 'toUpperCase', 'charAt', 'charCodeAt', 'concat', 'startsWith', 'endsWith', 'includes', 'match'),
            numberToString: (a, radix) => ctx => Number(a(ctx)).toString(radix(ctx)),
            tMap,
            tReduce
        }

        
        function resolveArgs(argListIndex: number) {
            return def.argLists[argListIndex].map(({value}) => 
                resolveSlot(def.slots[def.args[value].value]))
        }

        function resolveSlot(slot: Slot) : Evaluator {
            const args = resolveArgs(slot.args) 
            const op = resolvers[def.nativeOps[slot.op]]
            return op(...args)
        }

        function merge(table: number, value: any) {
            return {...tables[table], ...value}
        }

        function assign(table: number, value: any) {
            tables[table] = value
        }

        function evaluateMemoized(slot: number, ctx: Context | null = null) {
            if (ctx)
                return slots[slot](ctx)

            if (Reflect.has(memos, slot))
                return memos[slot]

            return (memos[slot] = slots[slot](null))
        }

        function enqueue(port: number, buffer: ArrayBuffer) {
            assign(def.locations.inbox, merge(def.locations.inbox, {[nextID()]: {port, buffer}}))
            cycle()
        }

        function cycle() {
            for (;;) {
                const assignments = evaluateMemoized(def.locations.staging) as {[table: number]: any}
                for (let table in assignments)
                    assign(+table, assignments[+table])

                memos = {}

                const nextWakeupTime = evaluateMemoized(def.locations.nextWakeupTime)
                const outbox = evaluateMemoized(def.locations.outbox)
                for (let port in outbox)
                    dequeue(+port, outbox[port])

                const now = Date.now()
                if (nextWakeupTime <= now) 
                    continue

                setTimeout(cycle, nextWakeupTime - now)
                break
            }
        }

        cycle()
        return enqueue
    }    
}
