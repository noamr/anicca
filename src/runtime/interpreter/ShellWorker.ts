import createShell from '../common/Shell'
import { StoreSpec } from '../../builder/types'
import createStoreInterpreter from './StoreInterpreter'
import { StartMessage } from '../common/RuntimeTypes'

export default function initInterpreterWorker(spec: StoreSpec) {
    const store = createStoreInterpreter(spec)
    self.onmessage = ev => {
        if (ev.data.type === 'start') {
            const {incomingPorts, outgoingPorts} = ev.data.payload as StartMessage
            createShell({incomingPorts, outgoingPorts, store})
        }
    }
}
