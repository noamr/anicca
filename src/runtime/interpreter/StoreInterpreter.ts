import {RawFormula, StoreSpec} from '../../builder/types'
import { Enqueue, Store } from '../shell/RuntimeTypes'
import {defaultEvaluators, Context, Evaluator} from './evaluators'

const noopSymbol = Symbol('noop')
const deleteSymbol = Symbol('delete')
const replaceSymbol = Symbol('replace')
const mergeSymbol = Symbol('merge')



export default function createStoreInterpreter(spec: StoreSpec) {
    const tables: {[index: number]: Map<any, any>} = {}
    const nextID = ((n: number) => () => ++n)(0)

    const ops: {[op: string]: (...args: Evaluator[]) => Evaluator} = {
        ...defaultEvaluators,
        table: (a) => ctx => tables[a(ctx)],
        uid: () => nextID,
        delete: (t, k) => (c) => new Map([[0, t(c)], [1, k(c)], [2, deleteSymbol]]),
        put: (t, k, v) => (c) => new Map([[0, t(c)], [1, k(c)], [2, v(c)]]),
        replace: () => () => replaceSymbol,
        merge: () => () => mergeSymbol,
        noop: () => () => noopSymbol
    }

    Object.keys(spec.tableTypes).forEach((tableIndex: string) => {
        tables[+tableIndex] = new Map()
    })

    const evaluators = new Map<number, Evaluator>()

    function resolveFormula(f: RawFormula): Evaluator {
        if (Reflect.has(f, 'value')) {
            const value = Reflect.get(f, 'value')
            return () => value
        }

        const {op, args} = f as {op: string, args: number[]}
        const ev = ops[op]
        const argResolvers = args.map(n => {
            if (!evaluators.has(n))
                evaluators.set(n, resolveFormula(spec.slots[n]))
            return evaluators.get(n) as Evaluator
        })

        return ev(...argResolvers)
    }

    function evaluate<T = any>(index: number, ctx: Context|null = null): T {
        return (evaluators.get(index) as Evaluator)(ctx)
    }

    function update(table: number, key: any, value: any) {
        const ensure = () => (tables[table] || (tables[table] = new Map<any, any>())) as Map<any, any>
        switch (key) {
            case noopSymbol:
                return

            case replaceSymbol:
                tables[table] = new Map(value as Map<any, any>)
                break

            case mergeSymbol:
                [...value].forEach(([k, v]) => ensure().set(k, v))
                break

            default:
                if (value === deleteSymbol)
                    ensure().delete(key)
                else
                    ensure().set(key, value)
        }
    }

    function commitEntry(entry: Map<number, number>) {
        const [table, key, value] = [entry.get(0), entry.get(1), entry.get(2)]
        update(table as number, key, value)
    }

    return {
        enqueue: (header: number, payload: ArrayBuffer|null) =>
            update(evaluate(spec.roots.inbox), nextID(), [header, payload]),
        awaitIdle: async () => evaluate<boolean>(spec.roots.idle),
        commit: async () => {
            const stagingData = await evaluate<Map<number, Map<number, number>>>(spec.roots.staging)
            for (const entry of stagingData.values())
                commitEntry(entry)
        },
        dequeue: async (): Promise<Array<[number, ArrayBuffer]>> => {
            const outbox = evaluate<Map<number, ArrayBuffer>>(spec.roots.outbox)
            const commitView = await evaluate<Map<number, number>>(spec.roots.commitView)
            commitEntry(commitView)
            return [...outbox.entries()]
        }
    } as Store
}
