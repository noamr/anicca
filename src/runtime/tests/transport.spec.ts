import { decode, encode } from '../interpreter/transport'
import { NativeType } from '../../builder/types'

const testEncodeDecode = (value: any, type: NativeType, fuzz: number|null = null) => it(JSON.stringify({value, type}), () => {
    const encoded = encode(value, type)
    expect(new Uint8Array(encoded).toString()).toMatchSnapshot()
    const decoded = decode(encoded, type)
    if (fuzz)
        expect(decoded).toBeCloseTo(value)
    else 
        expect(decoded).toEqual(value)
})
describe('transports', () => {
    testEncodeDecode(8, 'u8')
    testEncodeDecode(800, 'u16')
    testEncodeDecode(-8, 'i8')
    testEncodeDecode(-2342, 'i16')
    testEncodeDecode(0xFF34, 'u32')
    testEncodeDecode(-800000000, 'i32')
    testEncodeDecode(0xFF34FF, 'u64')
    testEncodeDecode(-8000000000000, 'i64')
    testEncodeDecode(1238.0333, 'f32', .003)
    testEncodeDecode(12383242.03333, 'f64')
    testEncodeDecode('hello world', 'string')
    testEncodeDecode('', 'string')
    testEncodeDecode(['abc', 123, 'def'], {tuple: ['string', 'u8', 'string']})
    testEncodeDecode(['abc', 123, [456, ['def']]], {tuple: ['string', 'u8', {tuple: ['u64', {tuple: ['string']}]}]})
    testEncodeDecode(new Map([[0, 'a'], [123, 'bla bla']]), {dictionary: ['u16', 'string']})
    testEncodeDecode(new Map([[0, ['a', new Map([[true, ['yes']]])]], [123, ['bla bla', new Map([[true, ['ok']], [false, ['not ok']]])]]]), 
                    {dictionary: ['f32', {tuple: ['string', {dictionary: ['bool', {tuple: ['string']}]}]}]})
    testEncodeDecode(true, 'bool')
})
