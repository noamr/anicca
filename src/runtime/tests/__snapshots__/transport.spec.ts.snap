// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transports {"value":"","type":"string"} 1`] = `"0,0,0,0"`;

exports[`transports {"value":"hello world","type":"string"} 1`] = `"0,0,0,11,104,101,108,108,111,32,119,111,114,108,100"`;

exports[`transports {"value":["abc",123,"def"],"type":{"tuple":["string","u8","string"]}} 1`] = `"0,0,0,3,97,98,99,123,0,0,0,3,100,101,102"`;

exports[`transports {"value":["abc",123,[456,["def"]]],"type":{"tuple":["string","u8",{"tuple":["u64",{"tuple":["string"]}]}]}} 1`] = `"0,0,0,3,97,98,99,123,64,124,128,0,0,0,0,0,0,0,0,3,100,101,102"`;

exports[`transports {"value":{},"type":{"dictionary":["f32",{"tuple":["string",{"dictionary":["bool",{"tuple":["string"]}]}]}]}} 1`] = `"0,0,0,2,0,0,0,0,0,0,0,1,97,0,0,0,1,1,0,0,0,3,121,101,115,66,246,0,0,0,0,0,7,98,108,97,32,98,108,97,0,0,0,2,1,0,0,0,2,111,107,0,0,0,0,6,110,111,116,32,111,107"`;

exports[`transports {"value":{},"type":{"dictionary":["u16","string"]}} 1`] = `"0,0,0,2,0,0,0,0,0,1,97,0,123,0,0,0,7,98,108,97,32,98,108,97"`;

exports[`transports {"value":-8,"type":"i8"} 1`] = `"248"`;

exports[`transports {"value":-2342,"type":"i16"} 1`] = `"246,218"`;

exports[`transports {"value":-800000000,"type":"i32"} 1`] = `"208,80,248,0"`;

exports[`transports {"value":-8000000000000,"type":"i64"} 1`] = `"194,157,26,148,162,0,0,0"`;

exports[`transports {"value":8,"type":"u8"} 1`] = `"8"`;

exports[`transports {"value":800,"type":"u16"} 1`] = `"3,32"`;

exports[`transports {"value":1238.0333,"type":"f32"} 1`] = `"68,154,193,17"`;

exports[`transports {"value":65332,"type":"u32"} 1`] = `"0,0,255,52"`;

exports[`transports {"value":12383242.03333,"type":"f64"} 1`] = `"65,103,158,129,65,17,10,19"`;

exports[`transports {"value":16725247,"type":"u64"} 1`] = `"65,111,230,159,224,0,0,0"`;

exports[`transports {"value":true,"type":"bool"} 1`] = `"1"`;
